core-bundle/src/Twig/Interop/ContextFactory.php

```diff
@@ -0,0 +1,175 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Twig\Interop;
+
+use Contao\Template;
+
+/**
+ * @experimental
+ */
+final class ContextFactory
+{
+    /**
+     * Create a Twig template context from a @see Template object.
+     */
+    public function fromContaoTemplate(Template $template): array
+    {
+        $context = $template->getData();
+
+        array_walk_recursive(
+            $context,
+            function (&$value, $key): void {
+                if ($value instanceof \Closure) {
+                    $value = $this->getCallableWrapper($value, (string) $key);
+                }
+            }
+        );
+
+        if (!isset($context['Template'])) {
+            $context['Template'] = $template;
+        }
+
+        return $context;
+    }
+
+    /**
+     * Create a Twig template context from an arbitrary object. This will also
+     * make protected methods/properties/constants accessible.
+     */
+    public function fromClass(object $object): array
+    {
+        $class = new \ReflectionClass($object);
+
+        $context = iterator_to_array($this->getAllMembers($object));
+
+        foreach ($class->getReflectionConstants() as $constant) {
+            if ($constant->isPrivate()) {
+                continue;
+            }
+
+            $context[$constant->getName()] = $constant->getValue();
+        }
+
+        foreach ($class->getStaticProperties() as $property => $value) {
+            if ($class->getProperty($property)->isPrivate()) {
+                continue;
+            }
+
+            $context[$property] = $value;
+        }
+
+        foreach ($class->getMethods() as $method) {
+            if ($method->isPrivate()) {
+                continue;
+            }
+
+            $name = $method->getName();
+
+            if (0 === strpos($name, '__')) {
+                continue;
+            }
+
+            $context[$name] = $this->getCallableWrapper($method->getClosure($object), $name);
+        }
+
+        if (!isset($context['data'])) {
+            $context['data'] = $object;
+        }
```

Comment:
This could easily result in naming conflicts I would guess. Can we use a more uncommon (for template variables) name. How about using `this`?

Reply:
Agreed. Changed in 98419a429.